### ISSUE TITLE
Pods get scheduled to nodes with insufficient space when using Inline Ephemeral Volumes

### ISSUE THREAD

**Issue by gtsystem** [topolvm/topolvm#117]
**Describe the bug**
When multiple pods using  inline ephemeral volume get created at the same time, some of them may get scheduled in nodes with insufficient space.
The reason for that is that the scheduler will call the extender for each pod with stale node information.

**Environments**
- Version: topolvm 0.4.0, k8s 1.16.3
- OS: linux

**To Reproduce**
Create a deployment as follow:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: topolvm-eph
spec:
  replicas: 5
  selector:
    matchLabels:
      app: topolvm-eph
  template:
    metadata:
      labels:
        app: topolvm-eph
    spec:
      containers:
      - name: ubuntu
        image: quay.io/cybozu/ubuntu:18.04
        command: ["/usr/local/bin/pause"]
        volumeMounts:
        - mountPath: /test
          name: my-volume
      volumes:
        - name: my-volume
          csi:
            driver: topolvm.cybozu.com
            fsType: xfs
            volumeAttributes:
              topolvm.cybozu.com/size: "3"
```

In my case there is a single node with 10GiB and each pod is asking for 3GiB. 

    kubectl create -f examples/pure-topolvm/dep-eph.yaml 

**Expected behavior**
Since there are 5 pods created, we expect only 3 of them can run on the node and 2 should stay in state pending.

**Actual behavior**
```
$ kubectl get pods
NAME                           READY   STATUS              RESTARTS   AGE
topolvm-eph-67d784cdb4-8fzx7   0/1     ContainerCreating   0          55m
topolvm-eph-67d784cdb4-k7hz4   1/1     Running             0          55m
topolvm-eph-67d784cdb4-kccmh   1/1     Running             0          55m
topolvm-eph-67d784cdb4-rthnf   1/1     Running             0          55m
topolvm-eph-67d784cdb4-xggln   0/1     ContainerCreating   0          55m
```

**Comment by morimoto-cybozu** [comment]
Hi @gtsystem , thanks for reporting the problem.

It's true that creating multiple pods simultaneously could confuse pod scheduling.
I discussed this issue with other maintainers and would like to leave it as a current limitation.
I will note it on the document.

This type of problem has been discussed in the k8s community.
The following open KEP is to extend the k8s Pod scheduler.
  [Storage Capacity Constraints for Pod Scheduling](https://github.com/kubernetes/enhancements/pull/1353)
The following issue is about the descheduling of failed-to-start pods.
  [deschedule pods that fail to start or restart too often](https://github.com/kubernetes-sigs/descheduler/issues/62)
IMO the assignment of node storage should be handled centrally by the k8s core scheduler.

I suppose that pod scheduling is also affected by CPU and memory assignment and pods will be distributed appropriately in many cases.
If you have experienced this type of problem in a real environment, would you please tell us your use case?

**Comment by gtsystem** [comment]
No, we are not running this setup in production at the moment, I just did this experiment as I was under the impression that the current scheduler extender is not able to cope with this issue.
I believe that the new scheduling framework will allow (when released) to solve this issue.

A custom "reserve" logic can be used to keep track of the used space in the node: [20180409-scheduling-framework.md](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/20180409-scheduling-framework.md#reserve)

**Comment by morimoto-cybozu** [comment]
Oh, I didn't know that KEP.  It indeed seems to solve the present problem.
I'll keep watching its implementation progress.
Thank you for the information!

**Comment by toshipp** [comment]
This is a limitation. closed via  #119
